docs/tutorial/part_1.rst

```diff
@@ -0,0 +1,238 @@
+
+        def __str__(self):
+            return f"Room({self.name} {self.host})"
+        
+    class Message(models.Model):
+        room = models.ForeignKey("chat.Room", on_delete=models.CASCADE, related_name="messages")
+        text = models.TextField(max_length=500)
+        user = models.ForeignKey(User, on_delete=models.CASCADE, related_name="messages")
+
+        def __str__(self):
+            return f"Message({self.user} {self.room})"
+        
+Creating the Serializers
+------------------------
+
+We will put the following code in the ``serializers.py`` file, to handle the serialization of the models created.
+
+.. code-block:: python
+
+    from .models import User, Room, Message
+    from rest_framework import serializers
+
+    class UserSerializer(serializers.ModelSerializer):
+        class Meta:
+            model = User
+            exclude = ["password"]
+
+    class MessageSerializer(serializers.ModelSerializer):
+        created_at_formatted = serializers.SerializerMethodField()
+        user = UserSerializer()
+        class Meta:
+            model = Message
+            exclude = []
+            depth = 1
+
+        def get_created_at_formatted(self, obj:Message):
+            return obj.created_at.strftime("%d-%m-%Y %H:%M:%S")
+
+    class RoomSerializer(serializers.ModelSerializer):
+        last_message = serializers.SerializerMethodField()
+        messages = MessageSerializer(many=True, read_only=True)
+        class Meta:
+            model = Room
+            fields = ["pk", "name", "host", "messages", "current_users", "last_message"]
+            depth = 1
+            read_only_fields = ["messages", "last_message"]
+            
+        def get_last_message(self, obj:Room):
+            return MessageSerializer(obj.messages.order_by('created_at').last()).data
+
+
+Creating the Consumers
+----------------------
+
+In the ``consumers.py`` file we will create only the room consumer for:
+    * Joining and leaving a room.
+    * Observing messages in that room.
+    * Observing the current users in the room.
+
+.. code-block:: python
+
+    import json
+    from django.shortcuts import get_object_or_404
+    from channels.generic.websocket import AsyncWebsocketConsumer
+    from channels.db import database_sync_to_async
+    from django.utils.timezone import now
+    from django.conf import settings
+    from typing import Generator
+    from djangochannelsrestframework.generics import GenericAsyncAPIConsumer, AsyncAPIConsumer
+    from djangochannelsrestframework.observer.generics import (ObserverModelInstanceMixin, action)
+    from djangochannelsrestframework.observer import model_observer
+
+    from .models import Room, Message, User
+    from .serializers import MessageSerializer, RoomSerializer, UserSerializer
+
+    class RoomConsumer(ObserverModelInstanceMixin, GenericAsyncAPIConsumer):
+        queryset = Room.objects.all()
+        serializer_class = RoomSerializer
+        lookup_field = "pk"
+
```

Comment:
since the users in a room is persisted in the db what about using this info on connect to auto subscribe the user to the rooms they are in (eg if they connect to the server from another device/tab)

Reply:
The users in a room are deleted when the user leaves the chat page (or that is how I was going to implement it), but is a good idea.

Reply:
The tracking of this in the DB might well make the tutorial more complex than it needs to be.

## A more complex extended stage 3 (maybe for anther PR) tutorial would be a good place for DB based tracking:

For tracking current_users in each room in the DB: I would suggest having the consumer subscribe (on connect) to rooms based on if the user is in the room. Then users can be added/removed from rooms from different places (eg an admin could kick them out) and the consumer would then un-subscribe the user from messages in that room. (does that make sense)